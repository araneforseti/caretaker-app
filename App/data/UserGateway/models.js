const Permissions = {
    FOCUS: 'focus',
    ADMIN: 'admin',
    HELPER: 'helper'
}

const NotificationTypes = {
    PUSH: 'push',
    EMAIL: 'email',
    SMS: 'sms'
}

const NotificationTriggers = {
    ANNOUNCEMENT: 'announcement',
    EVENT: 'event',
    EVENT_BY_ROLE: 'event by role'
}

const Settings = {
    [NotificationTypes.PUSH]: {
        name: 'Push',
        value: true
    },
    [NotificationTypes.EMAIL]: {
        name: 'Email',
        value: false
    },
    [NotificationTypes.SMS]: {
        name: 'SMS',
        value: false
    },
    [NotificationTriggers.ANNOUNCEMENT]: {
        name: 'Announcements',
        value: true
    },
    [NotificationTriggers.EVENT]: {
        name: 'Events',
        value: false
    },
    [NotificationTriggers.EVENT_BY_ROLE]: {
        name: 'Events specific to your role',
        value: true
    }
}

module.exports = {
  Permissions: Permissions,
  NotificationTypes: NotificationTypes,
  NotificationTriggers: NotificationTriggers,
  Settings: Settings
}
