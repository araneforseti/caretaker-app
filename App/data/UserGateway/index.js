import { Record } from 'immutable';
import UserModels from './models.js'
import { API_URL } from '../../config/ApiConfig';

export const Permissions = UserModels.Permissions
export const NotificationTypes = UserModels.NotificationTypes
export const NotificationTriggers = UserModels.NotificationTriggers

const createUserSettings = () => ({
  [NotificationTypes.PUSH]: {
    name: 'Push',
    value: true
  },
  [NotificationTypes.EMAIL]: {
    name: 'Email',
    value: false
  },
  [NotificationTypes.SMS]: {
    name: 'SMS',
    value: false
  },
  [NotificationTriggers.ANNOUNCEMENT]: {
    name: 'Announcements',
    value: true
  },
  [NotificationTriggers.EVENT]: {
    name: 'Events',
    value: false
  },
  [NotificationTriggers.EVENT_BY_ROLE]: {
    name: 'Events specific to your role',
    value: true
  }
})

export const User = Record({
  name: 'Defaul User',
  identifier: 'default@domain.com',
  permission: Permissions.HELPER,
  roles: [],
  phone: '123-456-7890',
  settings: createUserSettings()
});

const createUserGateway = function() {
  let allUsers = null;
  let currentUser = null;

  return {
    getCurrentUser() {
      return currentUser;
    },

    getAll() {
      if (allUsers === null) {
        //api call goes here
        allUsers = [
          new User({ name: 'S', identifier: 's', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend'], settings: createUserSettings() }),
          new User({ name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone: '123-456-7890', settings: createUserSettings() })
        ]
      }
      return allUsers;
    },

    getName(identifier) {
      const user = allUsers.find((user) => user.identifier === identifier)
      return user ? user.name : 'TBD';
    },

    getRoles() {
      return currentUser.roles
    },

    login(email) {
      //make call to google oauth and store the results
      let results = {
        email: email
      };
      let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
      if (typeof userFromBackend !== 'undefined') {
        currentUser = userFromBackend;
        return true;
      }
      return false;
    },

    logout() {
      // make call to oauth and/or invalidate current session
      this.currentUser = null;
    },

    isSelf(identifier) {
      return currentUser.identifier === identifier;
    },

    isAdmin() {
      return currentUser.permission === Permissions.ADMIN ||
        currentUser.permission === Permissions.FOCUS;
    },

    getCurrentPermissions() {
      return currentUser.permission;
    },

    inviteUser(email) {
      //api call to invite the user
      alert('invited ' + email + '!');
    },

    updateSettings(settings) {
      //api call to update the user settings
    }
  }
}

export default createUserGateway();
