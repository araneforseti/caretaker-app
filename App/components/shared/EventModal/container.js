import React, {Component} from 'react';
import UserGateway from '../../../data/UserGateway';
import EventGateway from '../../../data/EventGateway';

export var ContainerFor = (EventModal) => class extends Component {

    _onPickUpEvent = () => {
        EventGateway.pickUpEvent(this.props.event.id, UserGateway.getCurrentUser().identifier);
    };

    _onDropEvent = () => {
        EventGateway.dropEvent(this.props.event.id);
    }

    _onEditEvent = () => alert('edited!');

    _onExportToCalendar = () => alert('exported!');

    isVolunteerNeeded() {
        return this.props.event.volunteerId === 'TBD';
    }

    isVolunteerSelf() {
        return UserGateway.isSelf(this.props.event.volunteerId);
    }

    isVolunteerAdmin() {
        return UserGateway.isAdmin(this.props.event.volunteerId);
    }

    render() {
        if(this.props.event === null) {
            return null;
        }
        return (
            <EventModal
                visible={this.props.visible}
                event={this.props.event}
                isVolunteerNeeded={this.isVolunteerNeeded()}
                isVolunteerSelf={this.isVolunteerSelf()}
                isVolunteerAdmin={this.isVolunteerAdmin()}
                onPickUpEvent={this._onPickUpEvent}
                onDropEvent={this._onDropEvent}
                onEditEvent={this._onEditEvent}
                onExportToCalendar={this._onExportToCalendar}
                onClose={this.props.onClose}
            />
        )
    }
}
