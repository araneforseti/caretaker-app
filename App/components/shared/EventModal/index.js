import React, {Component} from 'react';
import {
    Button,
    FlatList,
    Modal,
    StyleSheet,
    Text,
    TouchableOpacity,
    View
} from 'react-native';
import UserGateway from '../../../gateway/user.js';

export default class EventModal extends Component {
    _renderButton = (text, onPress) => (
        <TouchableOpacity onPress={onPress}>
            <View style={styles.button}>
                <Text>{text}</Text>
            </View>
        </TouchableOpacity>
    );

    _onPickUpEvent = () => alert('volunteered!');

    _onDropEvent = () => alert('un-volunteered :(');

    _onEditEvent = () => alert('edited!');

    _onExportToCalendar = () => alert('exported!');

    _renderPickUpEventbutton = (volunteer) => {
        if(volunteer === 'TBD') {
            return this._renderButton('[pick up event]', this._onPickUpEvent);
        } else {
            return null;
        }
    }

    _renderDropEventButton = (volunteerId) => {
        if(UserGateway.isSelf(volunteerId)) {
            return this._renderButton('[drop event]', this._onDropEvent);
        } else {
            return null;
        }
    }

    _renderEditEventButton = (volunteerId) => {
        if(UserGateway.isAdmin(volunteerId)) {
            return this._renderButton('[edit event]', this._onEditEvent);
        } else {
            return null;
        }
    }

    _renderVolunteerInfo = (role, volunteer, volunteerId) => {
        return (
            <View>
                <Text style={styles.modalInfo}>{role}:{volunteer}</Text>
                {this._renderPickUpEventbutton(volunteerId)}
                {this._renderDropEventButton(volunteerId)}
            </View>
        );
    }

    _renderModalContent = () => {
        const props = this.props;
        const event = props.event;
        return (
            <View style={styles.modalContent}>
            {this._renderEditEventButton(event.volunteer)}
            {this._renderButton('[Export to Calendar]', this._onExportToCalendar)}
            <Text style={styles.modalTitle}>{event.name}</Text>

            <Text style={styles.modalInfo}>{event.date} {event.time}</Text>
            <Text style={styles.modalInfo}>{event.location}</Text>

            {this._renderVolunteerInfo(event.role, event.volunteer, event.volunteerId)}

            <Text style={styles.modalInfo}>{event.description}</Text>
            {this._renderButton('Close', props.close)}
            </View>
        );
    };

    render() {
        if(this.props.event === null) {
            return null;
        }
        return (
            <Modal visible={this.props.visible}
                animationType={"fade"}
                transparent={false}
                onRequestClose={() => {alert("Modal has been closed.")}}>
                {this._renderModalContent()}
            </Modal>
        );
  }
}

const styles = StyleSheet.create({
  modalContent: {
    margin: 13,
    justifyContent: 'center'
  },
  modalTitle: {
    fontSize: 24,
    textAlign: 'center',
    margin: 15
  },
  modalInfo: {
    fontSize: 20,
    textAlign: 'left',
    margin: 10
  }
});
