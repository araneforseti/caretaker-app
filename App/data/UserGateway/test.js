import UserGateway from './index.js';
import { Record } from 'immutable';

describe('UserGateway', () => {

    it('should return false when attempting to sign in with a non-existent user', () => {
        expect(UserGateway.login('TBD')).toBe(false);
    });

    it('should return true when sign in is successful', () => {
        expect(UserGateway.login('sarah@emailprovider.com')).toBe(true);
    });

    it('should assign TBD as a name when TBD is the identifier', () => {
        expect(UserGateway.getName('TBD')).toBe('TBD');
    });

    it('should get the current user as an immutable record', () => {
        UserGateway.login('s');

        const user =  UserGateway.getCurrentUser();

        //Record.isRecord is in the immutable@4.0.0 release candidate
        expect(user instanceof Record).toBe(true);
    })
});
