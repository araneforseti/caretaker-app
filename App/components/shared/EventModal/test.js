import React, {Component} from 'react';
import renderer from 'react-test-renderer';
import EventModal from './index.js';
import UserGateway from '../../../data/UserGateway';
import EventGateway from '../../../data/EventGateway';

describe('EventModal', () => {
    let modal;
    let modalInstance;
    let intiallyVisible;
    let event;
    let _onClose;

    beforeEach(() => {
        UserGateway.login('sarah@emailprovider.com'); //todo: mock/stub gateways
        EventGateway.getAll();//todo: mock/stub gateways

        intiallyVisible = false;
        event = {id: 1, date: 'July 1', time: '13:00-14:00', location: '6818 Austin Center Blvd, Austin, TX 78731', name: 'Chemotherapy', volunteerId: 'jack@coolwebsite.com', role: 'driver', description: ''};
        _onClose = jest.fn();

        modal = renderer
            .create(<EventModal
                        visible={intiallyVisible}
                        event={event}
                        onClose={_onClose} />);
        modalInstance = modal.getInstance();
    });

    it('should render correctly', () => {
        expect(modal).toMatchSnapshot();
    });

    it('should determine if a volunteer is needed', () => {
        event.volunteerId = 'TBD';

        expect(modalInstance.isVolunteerNeeded()).toBe(true);
    });

    it('should determine if a volunteer is not needed', () => {
        expect(modalInstance.isVolunteerNeeded()).toBe(false);
    });

    it('should check with the user gateway if the volunteer is the current user', () => {
        const spy = jest.spyOn(UserGateway, 'isSelf')
                        .mockReturnValue(true);

        expect(modalInstance.isVolunteerSelf()).toBe(true);
    });

    it('should check with user gateway if the volunteer is an admin', () => {
        const spy = jest.spyOn(UserGateway, 'isAdmin')
                        .mockReturnValue(true);

        expect(modalInstance.isVolunteerAdmin()).toBe(true);
    });

    xit('should not render if the event is not provided', () => {
        //todo: write this test in the parent component and move the functionality there
    });

    describe('_onPickUpEvent', () => {
        it('should update the event through the event gateway', () => {
            const spy = jest.spyOn(EventGateway, 'pickUpEvent');

            modalInstance._onPickUpEvent();

            expect(spy).toHaveBeenCalledWith(event.id, 'sarah@emailprovider.com');
        })
    });

    describe('_onDropEvent', () => {
        it('should update the event through the event gateway', () => {
            const spy = jest.spyOn(EventGateway, 'dropEvent');

            modalInstance._onDropEvent();

            expect(spy).toHaveBeenCalledWith(event.id);
        })
    });

    describe('_onEditEvent', () => {
        it('should alert because it hasnt really been implemented', () => {
            const spy = jest.spyOn(window, 'alert');

            modalInstance._onEditEvent();

            expect(spy).toHaveBeenCalledWith('edited!');
        });
    });

    describe('_onExportToCalendar', () => {
        it('should alert because it hasnt really been implemented', () => {
            const spy = jest.spyOn(window, 'alert');

            modalInstance._onExportToCalendar();

            expect(spy).toHaveBeenCalledWith('exported!');
        });
    });
});
