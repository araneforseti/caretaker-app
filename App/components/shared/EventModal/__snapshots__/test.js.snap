// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventModal should render correctly 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={false}
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ThemeProvider
      uiTheme={
        Object {
          "palette": Object {
            "accentColor": "#FF7575",
            "alternateTextColor": "white",
            "lightColor": "#FFC0C0",
            "primaryColor": "#FF7575",
            "primaryTextColor": "black",
            "secondaryColor": "#CC5D5D",
          },
          "spacing": Object {
            "miniActionButtonSize": 42,
          },
          "toolbar": Object {
            "container": Object {
              "height": 50,
            },
          },
          "typography": Object {
            "fontSize": Object {
              "display": 25,
            },
            "subheading": Object {
              "fontSize": 16,
              "fontWeight": "normal",
              "lineHeight": 24,
            },
          },
        }
      }
    />
    <RCTScrollView
      style={
        Object {
          "paddingTop": 22,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginBottom": 24,
            }
          }
        >
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
        </View>
        <ThemeProvider
          uiTheme={
            Object {
              "palette": Object {
                "accentColor": "#FF7575",
                "alternateTextColor": "white",
                "lightColor": "#FFC0C0",
                "primaryColor": "#FF7575",
                "primaryTextColor": "black",
                "secondaryColor": "#CC5D5D",
              },
              "spacing": Object {
                "miniActionButtonSize": 42,
              },
              "toolbar": Object {
                "container": Object {
                  "height": 50,
                },
              },
              "typography": Object {
                "fontSize": Object {
                  "display": 25,
                },
                "subheading": Object {
                  "fontSize": 16,
                  "fontWeight": "normal",
                  "lineHeight": 24,
                },
              },
            }
          }
        />
      </View>
    </RCTScrollView>
    <ThemeProvider
      uiTheme={
        Object {
          "palette": Object {
            "accentColor": "#FF7575",
            "alternateTextColor": "white",
            "lightColor": "#FFC0C0",
            "primaryColor": "#FF7575",
            "primaryTextColor": "black",
            "secondaryColor": "#CC5D5D",
          },
          "spacing": Object {
            "miniActionButtonSize": 42,
          },
          "toolbar": Object {
            "container": Object {
              "height": 50,
            },
          },
          "typography": Object {
            "fontSize": Object {
              "display": 25,
            },
            "subheading": Object {
              "fontSize": 16,
              "fontWeight": "normal",
              "lineHeight": 24,
            },
          },
        }
      }
    >
      <ActionButton
        icon="mode-edit"
        onPress={[Function]}
        style={
          Object {
            "container": Object {
              "backgroundColor": "#CC5D5D",
              "borderRadius": 21,
              "elevation": 4,
              "height": 42,
              "width": 42,
            },
            "positionContainer": Object {
              "left": 7,
              "position": "absolute",
              "top": 99,
            },
          }
        }
      />
    </ThemeProvider>
  </View>
</Modal>
`;
